overlord/snapstate/flags.go

```diff
@@ -50,3 +54,30 @@ func (f Flags) ForSnapSetup() Flags {
 	f.IgnoreValidation = false
 	return f
 }
+
+// EffectiveConfinement return the effective confinement for a given declared
+// confinement and flags like jailmode or devmode.
+func (f Flags) EffectiveConfinement(confinementType snap.ConfinementType) snap.ConfinementType {
+	switch confinementType {
+	case snap.ClassicConfinement:
+		// unchanged, flags don't impact snaps with classic confinement
+	case snap.DevmodeConfinement:
+		if f.JailMode {
+			// jailmode overrides devmode confinement declaration
+			return snap.StrictConfinement
+		}
+		// unchanged, stays devmode
+	case snap.StrictConfinement:
+		if f.JailMode {
+			// jailmode flag overrides devmode flag
+			return snap.StrictConfinement
+		}
```

Comment:
I think that having both flags (jail and dev mode) set is an edge case but I think that we should respect the jail mode flag just to err on the strict side. I don't think the code actually allows such a combination.

FYI: please let's continue to review this function in #2312 where it is coming from.